### PR TITLE
Use nuopc labels in fv3cap. Refactor ModelAdvance in fv3cap.

### PR DESCRIPTION
**Pull request by DusanJovic-NOAA**
## Description

* Use NUOPC semantic labels in fv3cap.
* Refactor ModelAdvance to call Model_Advance_phase1/2
* Remove ESMF_VERSION_MAJOR ifdef test. fv3atm now requires ESMF v8+

## Testing

Regression tests done on Hera (INtel and GNU)

## Dependencies

Do PRs in upstream repositories need to be merged first?
If so add the "waiting for other repos" label and list the upstream PRs

https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/113

### CODE REVIEW

File: fv3_cap.F90 [hunk 1]
```diff
@@ -102,125 +104,78 @@ subroutine SetServices(gcomp, rc)
...
-    end if
-    write(msgString,'(A,i6)') trim(subname)//' dbug = ',dbug
-    call ESMF_LogWrite(trim(msgString), ESMF_LOGMSG_INFO, rc=rc)
-
-  end subroutine
-
-  !-----------------------------------------------------------------------------
-
-  subroutine InitializeAdvertise(gcomp, importState, exportState, clock, rc)
+  subroutine InitializeAdvertise(gcomp, rc)
```
Comment (junwang-noaa) [inline fv3_cap.F90:165]: Is this standard interface? So if the importState, exportState and clock are updated within this subroutine, they can always be obtained through through NUOPC_ModelGet with gcomp?

Comment (DusanJovic-NOAA) [reply fv3_cap.F90:165]: According to NUOPC documentation:

```
 The user implemented specialization routines must follow the NUOPC interface definition.

  subroutine SpecRoutine(comp, rc)
    type(ESMF_*Comp)      :: comp
    integer, intent(out)  :: rc
  end subroutine
```
https://earthsystemmodeling.org/docs/release/ESMF_8_1_1/NUOPC_refdoc/node3.html#SECTION00031100000000000000

**Comment by DeniseWorthen** [comment]
Could we add this change to module_fcst_grid_comp.F90, which removes the requirement to have an empty data_table in the run directory. I've tested with control and 2 coupled configurations (one using data_override to apply runoff, one w/o) and all baselines pass.

```
diff --git a/module_fcst_grid_comp.F90 b/module_fcst_grid_comp.F90
index 618e86f..f9831b2 100644
--- a/module_fcst_grid_comp.F90
+++ b/module_fcst_grid_comp.F90
@@ -210,7 +210,7 @@ if (rc /= ESMF_SUCCESS) write(0,*) 'rc=',rc,__FILE__,__LINE__; if(ESMF_LogFoundE
     integer :: jsc, jec, isc, iec, nlev
     type(domain2D) :: domain
     integer :: n, fcstNpes, tmpvar
-    logical :: single_restart
+    logical :: single_restart, fexist
     integer, allocatable, dimension(:) :: isl, iel, jsl, jel
     integer, allocatable, dimension(:,:,:) :: deBlockList

@@ -404,9 +404,12 @@ if (rc /= ESMF_SUCCESS) write(0,*) 'rc=',rc,__FILE__,__LINE__; if(ESMF_LogFoundE
      call  atmos_model_init (atm_int_state%Atm,  atm_int_state%Time_init, &
                              atm_int_state%Time_atmos, atm_int_state%Time_step_atmos)
 !
-     call data_override_init ( ) ! Atm_domain_in  = Atm%domain, &
-                                 ! Ice_domain_in  = Ice%domain, &
-                                 ! Land_domain_in = Land%domain )
+     inquire(FILE='data_table', EXIST=fexist)
+     if (fexist) then
+       call data_override_init ( ) ! Atm_domain_in  = Atm%domain, &
+                                   ! Ice_domain_in  = Ice%domain, &
+                                   ! Land_domain_in = Land%domain )
+     endif
```

File: fv3_cap.F90 [hunk 2]
```diff
@@ -15,11 +15,13 @@ module fv3gfs_cap_mod
...
-                                    model_label_Advance     => label_Advance,     &
-                                    model_label_CheckImport => label_CheckImport, &
-                                    model_label_Finalize    => label_Finalize,    &
+  use NUOPC_Model,            only: model_routine_SS => SetServices,         &
+                                    routine_Run,                             &
+                                    label_Advertise,                         &
+                                    label_RealizeProvided,                   &
+                                    label_Advance,                           &
+                                    label_CheckImport,                       &
+                                    label_Finalize,                          &
```
Comment (rmontuoro) [inline fv3_cap.F90:25]: Add `label_TimestampExport`. See comment below.

File: fv3_cap.F90 [hunk 3]
```diff
@@ -99,125 +101,78 @@ subroutine SetServices(gcomp, rc)
...
                               specPhaseLabel="phase2", specRoutine=ModelAdvance_phase2, rc=rc)
     if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
 
     ! specializations required to support 'inline' run sequences
-    call NUOPC_CompSpecialize(gcomp, specLabel=model_label_CheckImport, &
+    call NUOPC_CompSpecialize(gcomp, specLabel=label_CheckImport, &
                               specPhaseLabel="phase1", specRoutine=NUOPC_NoOp, rc=rc)
     if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
 
     call NUOPC_CompSpecialize(gcomp, specLabel="ModelBase_TimestampExport", &
```
Comment (rmontuoro) [inline fv3_cap.F90]: Use `label_TimestampExport` instead of `"ModelBase_TimestampExport"`.

File: fv3_cap.F90 [hunk 4]
```diff
@@ -99,125 +101,78 @@ subroutine SetServices(gcomp, rc)
...
-    character(240)        :: msgString
-    logical               :: isPresent, isSet
-    character(len=*),parameter  :: subname='(fv3gfs_cap:InitializeP0)'
-
-    rc = ESMF_SUCCESS
-
-    ! Switch to IPDv01 by filtering all other phaseMap entries
-    call NUOPC_CompFilterPhaseMap(gcomp, ESMF_METHOD_INITIALIZE, &
-                                  acceptStringList=(/"IPDv01p"/), rc=rc)
-    if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
```
Comment (rmontuoro) [inline fv3_cap.F90:185]: Where is this statement going to end up?

File: fv3_cap.F90 [hunk 5]
```diff
@@ -99,125 +101,78 @@ subroutine SetServices(gcomp, rc)
...
-
-    call ESMF_AttributeGet(gcomp, name="ProfileMemory", value=value, defaultValue="true", &
```
Comment (rmontuoro) [inline fv3_cap.F90:187]: Are we removing the ability to set attributes for the FV3ATM component?

File: fv3_cap.F90 (hunk 4 above)
Comment (rsdunlapiv) [reply fv3_cap.F90:185]: With the NUOPC semantic labels, that statement is no longer needed.  No more IPDvXXpYY that the user needs to know about.  Hurrah!

File: fv3_cap.F90 (hunk 3 above)
Comment (DusanJovic-NOAA) [reply fv3_cap.F90]: Added in 49d660ff

File: fv3_cap.F90 (hunk 2 above)
Comment (DusanJovic-NOAA) [reply fv3_cap.F90:25]: Added in 49d660f

File: fv3_cap.F90 (hunk 4 above)
Comment (rmontuoro) [reply fv3_cap.F90:185]: Great!

File: fv3_cap.F90 (hunk 5 above)
Comment (junwang-noaa) [reply fv3_cap.F90:187]: see line 212

**Review by junwang-noaa — APPROVED** [review]

**Review by junwang-noaa — APPROVED** [review]

**Pull request merged**